core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -268,9 +274,32 @@ private void add(PendingDeleteFile file) {
     if (deleteFiles.add(file)) {
       addedFilesSummary.addedFile(spec, file);
       hasNewDeleteFiles = true;
+      if (ContentFileUtil.isDV(file)) {
+        newDVRefs.add(file.referencedDataFile());
+      }
+    }
+  }
+
+  protected void validateNewDeleteFile(DeleteFile file) {
+    Preconditions.checkNotNull(file, "Invalid delete file: null");
+    Preconditions.checkArgument(formatVersion() >= 2, "Deletes are supported in V2 and above");
+    if (file.content() == FileContent.POSITION_DELETES) {
+      Preconditions.checkArgument(
+          formatVersion() != 2 || !ContentFileUtil.isDV(file),
```

Comment:
This is a little more clear if we invert the check:

```suggestion
formatVersion() == 2 && ContentFileUtil.isDV(file),
```

Reply:
Wouldn't it fail for V3 tables, though?

Reply:
Yeah, you're right.  I feel like these are hard to read because of the way the checks stack on each other.